### ISSUE TITLE
Add support for more java.net.* classes

### ISSUE THREAD

**Issue by jvz** [pureconfig/pureconfig#325]
I'd primarily be interested in support `java.net.InetAddress` and `java.net.InetSocketAddress`.

Field type | Factory
-------|--------
`InetAddress` | `InetAddress::getByName`
`InetAddress[]` | `InetAddress::getAllByName`
`InetSocketAddress` | `InetSocketAddress::new` (may need to split on last `:` of string, or accept a map of `hostname` and `port` for construction)

Just for completion's sake, it may be nice to also support `NetworkInterface` as that can convert via a string (`NetworkInterface::getByName`) or an int (`NetworkInterface::getByIndex`), though I'm not sure how useful it would be in comparison.

**Comment by derekmorr** [comment]
We may run into issues with `InetAddress` since the JDK doesn't preserve inputs with IPv6. For example:

```scala
% scala
Welcome to Scala 2.12.4 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_151).
Type in expressions for evaluation. Or try :help.

scala> import java.net.InetAddress
import java.net.InetAddress

scala> InetAddress.getByName("2001:db8::123")
res0: java.net.InetAddress = /2001:db8:0:0:0:0:0:123

scala> res0.getCanonicalHostName
res1: String = 2001:db8:0:0:0:0:0:123
```
Note how `res` has extra zeros compared to the input. There are some third-party libraries that do a better job of this, but I don't know if it's worth pulling them in for this issue.

**Comment by jvz** [comment]
So the IPv6 host gets normalized to its proper IP? I mean, that sort of makes sense to me, though I'm not sure if that behavior would be surprising compared to all the other types that are supported.

**Comment by ruippeixotog** [comment]
I don't think it is an issue, it just means we have to be careful when adding tests for this. If `InetAddress` has a proper equals it should be no problem, since `2001:db8::123` and `2001:db8:0:0:0:0:0:123` are semantically equal IP addresses.

**Comment by ruippeixotog** [comment]
`InetAddress` and `InetSocketAddress` seem nice additions to pureconfig! I'm not very fond of reading `InetAddress[]` from a string though, since it makes pureconfig act differently than from collections of every other type... If there's demand for it maybe we can add it without `implicit` and make users import it explicitly, just like configurable converters.

**Comment by derekmorr** [comment]
Right, the issue is that IPv6 has multiple ways of representing the same address as text. I've added an `Arbitrary` instance to the test suite that supports IPv4 and IPv6. Several test runs show it working as expected.

**Comment by jvz** [comment]
The idea for using `Array[InetAddress]` isn't necessary an array, but some sort of collection class in general, though I'm not sure how that'd interact with the normal collection converters.

**Comment by ruippeixotog** [comment]
Yes, I also understood it that way. Even so, while it's easy to accept type with different string representations, I don't like the idea of reading a string to serialize to a list or to read a non-collection value to a collection by default (I have nothing against user-defined converters for that, though).

After looking into it a bit more, one thing troubles me even more: the fact that the converters you mention for `InetAddress` are not pure and actually depend on the network to actually work. While I'm perfectly ok with reading IP addresses as `InetAddress` (i.e. by using `getByAddress`), using `getByName` actually implies resolving a DNS name to an IP address, which can give different results at different times and actually depend on side-effects.

In those cases I'm really in favor of reading it as a hostname `String` and providing an `def ip: InetAddress` doing the conversion explicitly. pureconfig should only handle the reading of information fully contained in configs, not doing this kind of resolution. @derekmorr, what do you think?

**Comment by jvz** [comment]
Ah yes that is rather unfortunate. The main reason I even wanted this was for a more type-safe version of a hostname or hostname+port rather than passing around a string or string+int. Adding a `def` or `lazy val` or similar to my case classes is always an option, though now its internal state isn't really transparent like a typical case class.

**Comment by derekmorr** [comment]
There are some ways around this:

- the standard library's `InetAddress.getByName()` is pure if you pass in an IP address. It only does a DNS lookup if you pass in a hostname.

- As an alternative, there are third-party libraries that only support IP addresses, and will never do a DNS lookup. For example, Guava's InetAddresses can parse strings like "1.2.3.4" into a `java.net.InetAddress` - https://google.github.io/guava/releases/20.0/api/docs/com/google/common/net/InetAddresses.html Unfortunately Guava is a very heavyweight library, and I'm not sure it's worth adding a pureconfig module for it if we're only going to use a few types.

- If you just want to verify that a string has a vaild hostname, you can use Guava'a `InternetDomainName` - https://google.github.io/guava/releases/20.0/api/docs/com/google/common/net/InternetDomainName.html. That class has a pure function that just verifies that a string is structurally valid, but doesn't query DNS to see if it's nameserved. So, "foo.bar.org" is valid, but "foo!@#.-1.__org" is invalid.

- If you want to verify that an Int is a valid port number you can use refined. It has types for port numbers, system port numbers, user port numbers, etc - https://github.com/fthomas/refined/blob/master/modules/core/shared/src/main/scala/eu/timepit/refined/types/net.scala

I definitely think it's useful to have pureconfig support for `InetAddress`. I'm fine using the standard library's `InetAddress` helper functions as they're pure as long as you give the right input.

**Comment by ruippeixotog** [comment]
I agree that support for `InetAddress` would be useful and I'm perfectly ok with using pure constructors to create them. I'd also avoid including heavyweight libraries such as Guava as dependencies.

If the parsing isn't too complicated to implement, I'd parse IP addresses into bytes and use `getByAddress`, which is definitely pure. If parsing is too complicated but validating the input is not (i.e. with a regex), we can validate first and use `getByName` to parse, as you suggested.

**Comment by jvz** [comment]
A pureconfig-guava module would be far more appropriate for that, and there are many more types in that library which could potentially be useful I'm sure.

Parsing an IPv4 address is pretty trivial as a regexp, but IPv6 has the whole collapsed value thing going on which was pointed out as an impure side effect of creating an InetAddress from it.

**Comment by derekmorr** [comment]
IPv6 parsing is definitely more complicated b/c you can have embedded IPv4 addresses, or interface identifiers, etc. I'd rather not put that in pureconfig. So far, pureconfig delegates parsing and validation to other libraries.

We could look at a Guava module. I've already written `ConfigConvert` instances for `InetAddress` and `InternetDomainName` for work projects, so it's easy to factor that out into a pureconfig module.

**Comment by LeifW** [comment]
Would [InetSocketAddress.createUnresolved](https://docs.oracle.com/javase/7/docs/api/java/net/InetSocketAddress.html#createUnresolved(java.lang.String,%20int)) be an option for this?

**Comment by Philippus** [comment]
That's what I'm using in https://github.com/Philippus/pureconfig-javanet.
